erpnext/controllers/sales_and_purchase_return.py

```diff
@@ -101,8 +101,9 @@ def validate_returned_items(doc):
 								frappe.throw(_("Row # {0}: Serial No {1} does not match with {2} {3}")
 									.format(d.idx, s, doc.doctype, doc.return_against))
 
-				if warehouse_mandatory and not d.get("warehouse"):
-					frappe.throw(_("Warehouse is mandatory"))
+				if warehouse_mandatory and frappe.db.get_value("Item", d.item_code, "is_stock_item") \
```

Comment:
Can you check how `warehouse_mandatory` has been set?

I think it has an issue. It should be `warehouse_mandatory = not ((doc.doctype=="Purchase Invoice" or doc.doctype=="Sales Invoice") and not doc.update_stock))`